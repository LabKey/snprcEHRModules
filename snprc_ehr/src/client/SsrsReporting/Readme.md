## External reports are configured in the snprc_ehr.ExternalReports table using the format below.

- Id - identity field
- SortOrder - Optional integer value used to order report rows
- Label - display value in the report list
- Report - the SSRS path to the report and any optional hardcoded parameters e.g. *'CenterReports/IACUC/3YrRollover&Mode=1',*
- Description - report description
- Parameters - optional parameters 
```
    // parm format is: name, type, label, value
    // Each parm gets its own object
    // Example: 'TargetID:string:Target ID:31415, ...' = [{name: 'TargetId', type: 'string', label: 'Target ID', value: '31415' }, ...]
```    
- rsParameters - optional URL access parameters e.g., *'&rs:Command=Render'* 
[URL access parameter reference](https://docs.microsoft.com/en-us/sql/reporting-services/url-access-parameter-reference?view=sql-server-ver15)

## Source SSRS Server Configuration
The SSRSServerURL property is used to configure the reporting server viewer baseURL. To set its value use the *Folder>Management>Module Properties* UI.  
For example: `http://servername/ReportServer/Pages/ReportViewer.aspx?`

## Example table insert statement

```
	INSERT INTO snprc_ehr.ExternalReports
    (
        SortOrder, 
		Label,
        Report,
        Description,
        Parameters,
        rsParameters,
        Created,
        Modified,
        CreatedBy,
        ModifiedBy
    )
    VALUES
    (	7,
        N'IACUC 3yr Rollover (Standard)',       -- Label - nvarchar(64)
        N'CenterReports/IACUC/3YrRollover&Mode=1',       -- Value - nvarchar(400)
        N'Target Date is used to determine 3yr period.',       -- Description - nvarchar(4000)
        N'IACUC:string:IACUC, TargetDate:Date:Target Date',       -- Parameters - nvarchar(4000)
        N'&rc:Parameters=Collapsed',       -- rsParameters - nvarchar(4000)
        GETDATE(), -- Created - datetime
        GETDATE(), -- Modified - datetime
        (SELECT userId FROM core.Principals WHERE name LIKE '%thawkins%') ,      -- CreatedBy - USERID
        (SELECT userId FROM core.Principals WHERE name LIKE '%thawkins%')       -- ModifiedBy - USERID
    )
```