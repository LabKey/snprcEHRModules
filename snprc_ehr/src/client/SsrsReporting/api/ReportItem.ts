export interface ReportItem {
    id: number;
    label: string;
    value: string;
    description: string;
    parameters: string;
    rsParameters?:string;
}