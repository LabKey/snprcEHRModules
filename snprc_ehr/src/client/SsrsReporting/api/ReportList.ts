import { OptionsType } from 'react-select'

export interface ReportType {
    id: number;
    label: string;
    value: string;
    description: string;
    parameters: string;

}
export const reportList: OptionsType<ReportType> =  
    [
        {id: 1, label: 'Long Term Clinic Admissions', value: 'ReportS/report/LabkeyReports/LongTermClinic', description: 'Clinical admissions over 21 days',  parameters:'' },
        {id: 2, label: 'Overdue TB and Physical', value: 'ReportS/report/CenterReports/Routine%20Health/OverduePhyTB', description: 'Last TB test and physical exams that are over 180 days',  parameters:'' },
        {id: 3, label: 'Baboon weaning', value: 'ReportS/report/ColonyReports/Boon_weaning', description: 'Baboon infants that are ready to be weaned',  parameters:'' },
        {id: 4, label: 'Rhesus SPF weaning', value: 'ReportS/report/ColonyReports/rh2_weaning', description: 'Rhesus infants that are ready to be weaned', parameters:'' },
        {id: 5, label: 'Colony Census (current)', value: 'testReport1', description: 'Colony census as of today', parameters:'' },
        {id: 6, label: 'Colony Census (select date)', value: 'testReport2', description: 'Colony census on a specific date',  parameters:'targetDate:Date:Target Date' },
        {id: 7, label: 'Birth record report', value: 'ReportS/report/Development/BirthRecord', description: 'Birth certificate for animal',  parameters:'TargetID:string:Animal Id' },
        {id: 8, label: 'IACUC 3yr Rollover (Standard)', value: 'ReportS/report/CenterReports/IACUC/3YrRollover', description: 'Target Date is used to determine 3yr period NULL uses today.', parameters:'IACUC:string:IACUC, TargetDate:Date:Target Date, Mode:String:Mode:Standard' },
        {id: 8, label: 'IACUC 3yr Rollover (Backdate)', value: 'ReportS/report/CenterReports/IACUC/3YrRollover' ,description: 'Target Date sets the 3yr end date.  Start date is 3 years before target date.', parameters:'IACUC:string:IACUC, TargetDate:Date:Target Date, Mode:String:Mode:BackDate' },
        {id: 10, label: 'IACUC 3yr Rollover (Harddate)', value: 'ReportS/report/CenterReports/IACUC/3YrRollover', description: 'Target date sets the end date.  Start date is the date file start date before the target date. Could result in more or less than 3 yrs', parameters:'IACUC:string:IACUC, TargetDate:Date:Target Date, Mode:String:Mode:HardDate' }
    ];
    