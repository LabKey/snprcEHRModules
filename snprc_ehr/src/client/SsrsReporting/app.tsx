import React from 'react'
import ReactDOM from 'react-dom'

import SsrsReporting from './SsrsReporting'

// Need to wait for container element to be available in labkey wrapper before render
window.addEventListener('DOMContentLoaded', () => {
    ReactDOM.render(<SsrsReporting />, document.getElementById('app'))
})
