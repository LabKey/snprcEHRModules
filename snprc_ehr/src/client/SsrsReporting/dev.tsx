import React from 'react'
import ReactDOM from 'react-dom'
import { AppContainer } from 'react-hot-loader'

import SsrsReporting from './SsrsReporting'

const render = () => {
    ReactDOM.render(
      <AppContainer>
        <SsrsReporting />
      </AppContainer>,
        document.getElementById('app')
    )
}
declare const module: any;

if (module.hot) {
    module.hot.accept()
}

render()
