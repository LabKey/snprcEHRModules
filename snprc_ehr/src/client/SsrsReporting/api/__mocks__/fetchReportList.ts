import { OptionsType } from 'react-select'
import { ReportItem } from '../ReportItem'

// API test data
//Report server docs: https://docs.microsoft.com/en-us/sql/reporting-services/url-access-parameter-reference?view=sql-server-ver15
export const fetchReportList = () => {  
    return new Promise<OptionsType<ReportItem>> ( (resolve, reject) => {
        resolve(
            [
                {id: 1, label: 'Long Term Clinic Admissions', value: 'LabkeyReports/LongTermClinic', description: 'Clinical admissions over 21 days',  parameters:'', rsParameters:'&rs:Command=Render' },
                {id: 2, label: 'Overdue TB and Physical', value: 'CenterReports/Routine%20Health/OverduePhyTB', description: 'Last TB test and physical exams that are over 180 days',  parameters:'', rsParameters: '&rs:Command=Render' },
                {id: 3, label: 'Baboon weaning', value: 'ColonyReports/Boon_weaning', description: 'Baboon infants that are ready to be weaned',  parameters:'', rsParameters: '&rs:Command=Render' },
                {id: 4, label: 'Rhesus SPF weaning', value: 'ColonyReports/rh2_weaning', description: 'Rhesus infants that are ready to be weaned', parameters:'', rsParameters: '&rs:Command=Render' },
                {id: 5, label: 'Colony Census (current)', value: 'testReport1', description: 'Colony census as of today', parameters:'', rsParameters: '&rs:Command=Render' },
                {id: 6, label: 'Colony Census (select date)', value: 'testReport2', description: 'Colony census on a specific date',  parameters:'targetDate:Date:Target Date', rsParameters: '&rs:Command=Render' },
                {id: 7, label: 'Birth record report', value: 'Development/BirthRecord', description: 'Birth certificate for animal',  parameters:'TargetID:string:Animal Id', rsParameters: '&rc:Parameters=Collapsed' },
                {id: 8, label: 'IACUC 3yr Rollover (Standard)', value: 'CenterReports/IACUC/3YrRollover&Mode=1', description: 'Target Date is used to determine 3yr period.', parameters:'IACUC:string:IACUC, TargetDate:Date:Target Date', rsParameters: '&rc:Parameters=Collapsed' },
                {id: 8, label: 'IACUC 3yr Rollover (Backdate)', value: 'CenterReports/IACUC/3YrRollover&Mode=2' ,description: 'Target Date sets the end date.  Start date is 3 years before target date.', parameters:'IACUC:string:IACUC, TargetDate:Date:Target Date', rsParameters: '&rc:Parameters=Collapsed' },
                {id: 10, label: 'IACUC 3yr Rollover (Harddate)', value: 'CenterReports/IACUC/3YrRollover&Mode=3', description: 'Target date sets the end date.  Start date is the start date before the target date. Could result in more or less than 3 yrs.', parameters:'IACUC:string:IACUC, TargetDate:Date:Target Date', rsParameters: '&rc:Parameters=Collapsed' }
            ]
        ), (error: string ) => {
            reject(error)
        }
    
    })
}
    