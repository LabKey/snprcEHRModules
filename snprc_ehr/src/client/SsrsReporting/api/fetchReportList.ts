import { ReportItem } from './ReportItem'
import { request } from '../../Shared/api/api'

const parse = rows => {
    return rows.map( ( { data } ) => {
        const reportItem: ReportItem = 
         { 
            id: data.Id.value, 
            sortOrder: data.SortOrder.value,
            label: data.Label.value,
            value: data.Report.value,
            description: data.Description.value,
            parameters: data.Parameters.value,
            rsParameters: data.rsParameters.value
        }
        return reportItem;
    })
}

export const fetchReportList = (): Promise<ReportItem[]> => {
    return new Promise((resolve, reject) => {
        request({
            schemaName: 'snprc_ehr',
            queryName: 'ExternalReports',
            columns: ['Id', 'SortOrder', 'Label', 'Report', 'Description', 'Parameters', 'rsParameters'],
            sort: 'SortOrder'
        }).then(({ rows }) => {
            const parsedRows:ReportItem[] = parse(rows)
            resolve(parsedRows)
        }).catch(error => {
            console.log('error', error)
            reject(error)
        })
    })
}

    