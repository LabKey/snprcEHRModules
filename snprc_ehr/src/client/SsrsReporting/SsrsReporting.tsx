/* eslint-disable no-alert */

import React from "react";
import { LoadingSpinner } from "../Shared/components/LoadingSpinner";
import "./styles/ssrsReporting.scss";
import { Button } from "react-bootstrap";
import ReportSelectionPanel from "./components/ReportSelectionPanel";
import InfoPanel from "../Shared/components/InfoPanel";
import { OptionsType } from "react-select";
import { ReportItem } from "./api/ReportItem";
import { fetchReportList } from "./api/fetchReportList";
import ReportParmsPanel from "./components/ReportParmsPanel";
import {
    ReportParm,
    parseParms,
    parmsToQueryString,
} from "./services/parmUtils";
import { printToPDF } from "./services/printToPDF";

interface State {
    reportList: OptionsType<ReportItem>;
    selectedReport?: ReportItem;
    errorMessage: string;
    isLoading: boolean;
    reportParameters?: ReportParm[];
    parms: string;
}
interface Props {}

export default class SsrsReporting extends React.Component<Props, State> {
    constructor(props: Props) {
        super(props);

        this.state = {
            reportList: [],
            errorMessage: "",
            isLoading: true,
            parms: "",
        };
    }

    componentDidMount = async () => {
        await fetchReportList()
            .then((response: ReportItem[]) => {
                this.setState({ reportList: response, isLoading: false });
            })
            .catch((error) => {
                this.setState(() => ({
                    errorMessage: error.exception,
                    isLoading: false
                }));
            });
    };

    //quit
    onQuitClick = (): void => {
        window.history.back();
    };

    onClickPrint = () => {
        if (!this.state.selectedReport) {
            this.setState(() => ({
                errorMessage: "Select a report before clicking print.",
            }));
        } else {
            printToPDF(this.state.selectedReport, this.state.parms);
            this.setState(() => ({
                errorMessage: "",
            }));
        }
    };

    handleChange = (selectedReport: ReportItem) => {
        const parsedParms = parseParms(selectedReport.parameters);

        this.setState(
            () => ({
                errorMessage: "",
                selectedReport: selectedReport,
                reportParameters: parsedParms,
                parms: "",
            }),
            () => {}
        );
    };

    handleParmChange = (index: number, parmItem: ReportParm): void => {
        let parameters: ReportParm[] = [...this.state.reportParameters];

        parameters.splice(index, 1, parmItem);

        this.setState(() => ({
            reportParameters: parameters,
            parms: parmsToQueryString(parameters),
        }));
    };

    render() {
        const { isLoading } = this.state;

        if (isLoading) {
            return <LoadingSpinner msg="Loading app..." />;
        }

        return (
            <div>
                <div className="parent-panel">
                    <div className="report-div">
                        <ReportSelectionPanel
                            reportList={this.state.reportList}
                            handleChange={this.handleChange}
                            selectedReport={this.state.selectedReport}
                        />
                    </div>
                    <div className="panel-heading">
                        <p>Report Description and Parameters</p>
                    </div>
                    <div className="wizard-panel">
                        <ReportParmsPanel
                            reportParameters={this.state.reportParameters}
                            handleParmChange={this.handleParmChange}
                            selectedReport={this.state.selectedReport}
                        />
                    </div>
                    <InfoPanel
                        errorMessages={
                            this.state.errorMessage && [
                                {
                                    propTest: true,
                                    colName: this.state.errorMessage,
                                },
                            ]
                        }
                        infoMessages={[
                            {
                                key: 1,
                                value:
                                    "Select a report, enter parameters, and then press the print button",
                            },
                        ]}
                    />
                </div>

                <div className="wizard-footer">
                    <Button bsStyle="primary" onClick={this.onClickPrint}>
                        Print Report{" "}
                        <i aria-hidden="true" className="fa fa-print" />
                    </Button>
                    <Button onClick={this.onQuitClick}>Quit</Button>
                </div>
            </div>
        );
    }
}
