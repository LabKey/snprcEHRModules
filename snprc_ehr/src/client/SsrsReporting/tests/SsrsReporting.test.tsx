/* eslint-disable no-unused-vars */

import React from 'react'
import { shallow } from 'enzyme'
import SsrsReporting from '../SsrsReporting'

import { fetchReportList } from '../api/fetchReportList'
jest.mock('../api/fetchReportList')

require('setimmediate')

beforeAll(() => {
    // global.fetch = jest.fn();
})

let wrapper

beforeEach(() => {
    wrapper = shallow(<SsrsReporting />)
})

afterEach(() => {
    wrapper.unmount()
})

function flushPromises() {
    return new Promise(resolve => setImmediate(resolve))
}

describe('SsrsReporting tests', () => {

    test('Should render a loading spinner before rendering the landing page.', async () => {
        wrapper.unmount()
        wrapper = shallow(<SsrsReporting />, { disableLifecycleMethods: true })
        await flushPromises()
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('LoadingSpinner').exists()).toBeTruthy()
    })

    test('Should render the landing page.', async () => {
        await flushPromises()
        expect(wrapper).toMatchSnapshot()

        // spinner should be gone
        expect(wrapper.find('LoadingSpinner').exists()).toBeFalsy()
        // ReportSelectionPanel and ReportParmsPanel should be present
        expect(wrapper.find('ReportSelectionPanel').exists()).toBeTruthy()
        expect(wrapper.find('ReportParmsPanel').exists()).toBeTruthy()
    })

})
