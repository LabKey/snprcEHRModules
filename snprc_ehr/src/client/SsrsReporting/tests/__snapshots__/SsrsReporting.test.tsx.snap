// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SsrsReporting tests Should render a loading spinner before rendering the landing page. 1`] = `
<LoadingSpinner
  msg="Loading app..."
  wrapperClassName="loading-spinner"
/>
`;

exports[`SsrsReporting tests Should render the landing page. 1`] = `
<div>
  <div
    className="parent-panel"
  >
    <div
      className="report-div"
    >
      <ReportSelectionPanel
        handleChange={[Function]}
        reportList={
          Array [
            Object {
              "description": "Clinical admissions over 21 days",
              "id": 1,
              "label": "Long Term Clinic Admissions",
              "parameters": "",
              "rsParameters": "&rs:Command=Render",
              "value": "LabkeyReports/LongTermClinic",
            },
            Object {
              "description": "Last TB test and physical exams that are over 180 days",
              "id": 2,
              "label": "Overdue TB and Physical",
              "parameters": "",
              "rsParameters": "&rs:Command=Render",
              "value": "CenterReports/Routine%20Health/OverduePhyTB",
            },
            Object {
              "description": "Baboon infants that are ready to be weaned",
              "id": 3,
              "label": "Baboon weaning",
              "parameters": "",
              "rsParameters": "&rs:Command=Render",
              "value": "ColonyReports/Boon_weaning",
            },
            Object {
              "description": "Rhesus infants that are ready to be weaned",
              "id": 4,
              "label": "Rhesus SPF weaning",
              "parameters": "",
              "rsParameters": "&rs:Command=Render",
              "value": "ColonyReports/rh2_weaning",
            },
            Object {
              "description": "Colony census as of today",
              "id": 5,
              "label": "Colony Census (current)",
              "parameters": "",
              "rsParameters": "&rs:Command=Render",
              "value": "testReport1",
            },
            Object {
              "description": "Colony census on a specific date",
              "id": 6,
              "label": "Colony Census (select date)",
              "parameters": "targetDate:Date:Target Date",
              "rsParameters": "&rs:Command=Render",
              "value": "testReport2",
            },
            Object {
              "description": "Birth certificate for animal",
              "id": 7,
              "label": "Birth record report",
              "parameters": "TargetID:string:Animal Id",
              "rsParameters": "&rc:Parameters=Collapsed",
              "value": "Development/BirthRecord",
            },
            Object {
              "description": "Target Date is used to determine 3yr period.",
              "id": 8,
              "label": "IACUC 3yr Rollover (Standard)",
              "parameters": "IACUC:string:IACUC, TargetDate:Date:Target Date",
              "rsParameters": "&rc:Parameters=Collapsed",
              "value": "CenterReports/IACUC/3YrRollover&Mode=1",
            },
            Object {
              "description": "Target Date sets the end date.  Start date is 3 years before target date.",
              "id": 8,
              "label": "IACUC 3yr Rollover (Backdate)",
              "parameters": "IACUC:string:IACUC, TargetDate:Date:Target Date",
              "rsParameters": "&rc:Parameters=Collapsed",
              "value": "CenterReports/IACUC/3YrRollover&Mode=2",
            },
            Object {
              "description": "Target date sets the end date.  Start date is the start date before the target date. Could result in more or less than 3 yrs.",
              "id": 10,
              "label": "IACUC 3yr Rollover (Harddate)",
              "parameters": "IACUC:string:IACUC, TargetDate:Date:Target Date",
              "rsParameters": "&rc:Parameters=Collapsed",
              "value": "CenterReports/IACUC/3YrRollover&Mode=3",
            },
          ]
        }
      />
    </div>
    <div
      className="panel-heading"
    >
      <p>
        Report Description and Parameters
      </p>
    </div>
    <div
      className="wizard-panel"
    >
      <ReportParmsPanel
        handleParmChange={[Function]}
      />
    </div>
    <InfoPanel
      errorMessages=""
      infoMessages={
        Array [
          Object {
            "key": 1,
            "value": "Select a report, enter parameters, and then press the print button",
          },
        ]
      }
    />
  </div>
  <div
    className="wizard-footer"
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      disabled={false}
      onClick={[Function]}
    >
      Print Report
       
      <i
        aria-hidden="true"
        className="fa fa-print"
      />
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      disabled={false}
      onClick={[Function]}
    >
      Quit
    </Button>
  </div>
</div>
`;
